### DISCUSSION TITLE
How to tell vim the filetype of dot_bashrc (and other files) automatically?

### DISCUSSION THREAD

**Discussion by kaddkaka** [twpayne/chezmoi#2990] [Q&A]
The name mangling that chezmoi does prevents vim from by default figuring out what filetype my `.bashrc` file is when I open it inside my chezmoi dir because then vim just sees the path `.local/share/chezmoi/dot_bashrc`.

Has anyone else thought about this and solved it in a neat way?

I would like to avoid a modeline in the beginning of the file a la `# vi: ft=gitconfig` because I suspect there code be similar cases for other files as well.

**Comment by twpayne** [comment]
You have several options:

1. Use `chezmoi edit`. This will invoke your editor with a filename equal to the target filename, e.g. `.bashrc`.
2. Use [`chezmoi.vim`](https://github.com/alker0/chezmoi.vim).
3. Use a modeline. Many editors support these.

**Reply by halostatue** [reply]
I use `chezmoi.vim` and it works *very* nicely, but it is *very* fiddly on setup (it *must* run early).

What's nice is that it *can* combine Go template highlighting with the underlying filetype.

**Comment by Lockszmith-GH** [comment]
From my experience a [shebang](https://en.wikipedia.org/wiki/Shebang_(Unix)) is enough to allow vim to apply shell syntax highlighting.

When I add the following to the top of a file it works fine:
```
#! /usr/bin/env bash
```

If I want to prevent the script from running, and only work when sourced, I use the following:
```
#!/usr/bin/env -S bash -c 'echo "Not a user script. source(aka .) only"'
```

**Comment by kaddkaka** [comment]
Right, there are two reasons why I don't reach for `chezmoi edit`:
1. I'm often already in (n)vim when I realize I want to change a config file.
2. I tried to setup automatic apply when saving the config file, but my autocmd is a bit hacky, and not guaranteed to stay working (as per discussion here: https://github.com/twpayne/chezmoi/issues/2922#issuecomment-1506951332)
   ```vim
   autocmd BufWritePost /tmp/chezmoi-edit* ! chezmoi apply <afile>:s?/tmp/chezmoi-edit\d*/?~/?:S
   ```

My current Neovim config with regards to chezmoi looks like this and it works great apart from the filetype issue:

```vim
map <leader>e <cmd>Files $CHEZMOI_HOME<cr>
augroup init_group
  autocmd!
  autocmd BufWritePost ~/.local/share/chezmoi/* ! chezmoi apply --source-path <afile>
augroup END
```
